### PR TITLE
udp for bottom_up paradigm in 'switch' method

### PR DESCRIPTION
**Pull request by HuangJunJie2017**
udp for bottom_up paradigm in 'switch' method

### CODE REVIEW

File: mmpose/core/evaluation/bottom_up_eval.py [hunk 1]
```diff
@@ -11,7 +12,8 @@ def get_multi_stage_outputs(outputs,
                             tag_per_joint=True,
                             flip_index=None,
                             project2image=True,
-                            size_projected=None):
+                            size_projected=None,
+                            use_udp=False):
```
Comment (innerlee) [inline mmpose/core/evaluation/bottom_up_eval.py:16]: `use_udp` -> `align_corners`

File: mmpose/core/evaluation/bottom_up_eval.py [hunk 2]
```diff
@@ -102,22 +105,23 @@ def get_multi_stage_outputs(outputs,
...
+                align_corners=use_udp) for hms in heatmaps
         ]
 
         tags = [
             torch.nn.functional.interpolate(
                 tms,
                 size=(size_projected[1], size_projected[0]),
                 mode='bilinear',
-                align_corners=False) for tms in tags
+                align_corners=use_udp) for tms in tags
```
Comment (innerlee) [inline mmpose/core/evaluation/bottom_up_eval.py:116]: align_corners

File: mmpose/core/evaluation/top_down_eval.py [hunk 3]
```diff
@@ -250,6 +251,98 @@ def _taylor(heatmap, coord):
...
+    hessian[:, :, 0, 0] = dxx
+    hessian[:, :, 1, 0] = dxy
+    hessian[:, :, 0, 1] = dxy
+    hessian[:, :, 1, 1] = dyy
+    inv_hessian = np.zeros(hessian.shape)
+    for i in range(coord_shape[0]):
+        for j in range(coord_shape[1]):
+            hessian_tmp = hessian[i, j, :, :]
+            try:
+                inv_hessian[i, j, :, :] = np.linalg.inv(hessian_tmp)
```
Comment (innerlee) [inline mmpose/core/evaluation/top_down_eval.py:331]: can `hessian + eps * I`  avoid the inv issue?

File: mmpose/models/detectors/bottom_up.py [hunk 4]
```diff
@@ -35,7 +35,8 @@ def __init__(self,
                  train_cfg=None,
                  test_cfg=None,
                  pretrained=None,
-                 loss_pose=None):
+                 loss_pose=None,
+                 use_udp=False):
```
Comment (jin-s13) [inline mmpose/models/detectors/bottom_up.py:39]: Here, `use_udp` can be moved into `test_cfg`.

File: mmpose/models/detectors/bottom_up.py [hunk 5]
```diff
@@ -222,12 +224,13 @@ def forward_test(self, img, img_metas, return_heatmap=False, **kwargs):
                 outputs, outputs_flip, self.test_cfg['num_joints'],
                 self.test_cfg['with_heatmaps'], self.test_cfg['with_ae'],
                 self.test_cfg['tag_per_joint'], img_metas['flip_index'],
-                self.test_cfg['project2image'], base_size)
+                self.test_cfg['project2image'], base_size,
+                use_udp=self.use_udp)
```
Comment (jin-s13) [inline mmpose/models/detectors/bottom_up.py:228]: And, we can use `self.test_cfg['use_udp']`  here.

File: mmpose/core/post_processing/group.py [hunk 6]
```diff
@@ -141,11 +141,12 @@ def __init__(self, cfg):
 class HeatmapParser:
     """The heatmap parser for post processing."""
 
-    def __init__(self, cfg):
+    def __init__(self, cfg, use_udp=False):
         self.params = _Params(cfg)
         self.tag_per_joint = cfg['tag_per_joint']
         self.pool = torch.nn.MaxPool2d(cfg['nms_kernel'], 1,
                                        cfg['nms_padding'])
+        self.use_udp = use_udp
```
Comment (jin-s13) [inline mmpose/core/post_processing/group.py:149]: Here, `use_udp` can be moved into `cfg`

File: mmpose/models/detectors/bottom_up.py (hunk 5 above)
Comment (HuangJunJie2017) [reply mmpose/models/detectors/bottom_up.py:228]: done

File: mmpose/core/evaluation/bottom_up_eval.py (hunk 1 above)
Comment (HuangJunJie2017) [reply mmpose/core/evaluation/bottom_up_eval.py:16]: done

File: mmpose/core/evaluation/bottom_up_eval.py (hunk 2 above)
Comment (HuangJunJie2017) [reply mmpose/core/evaluation/bottom_up_eval.py:116]: done

File: mmpose/core/evaluation/top_down_eval.py (hunk 3 above)
Comment (HuangJunJie2017) [reply mmpose/core/evaluation/top_down_eval.py:331]: done, replace 'try' with 'hessian + eps * I '

File: mmpose/core/post_processing/group.py (hunk 6 above)
Comment (HuangJunJie2017) [reply mmpose/core/post_processing/group.py:149]: done

File: mmpose/core/post_processing/group.py [hunk 7]
```diff
@@ -273,7 +275,7 @@ def adjust(ans, heatmaps):
         return ans
 
     @staticmethod
-    def refine(heatmap, tag, keypoints):
+    def refine(heatmap, tag, keypoints, use_udp):
```
Comment (jin-s13) [inline mmpose/core/post_processing/group.py:278]: `use_udp=False`

Comment (HuangJunJie2017) [reply mmpose/core/post_processing/group.py:278]: done

**Comment by HuangJunJie2017** [comment]
emmm， how can i fix the remaining bug in bulid/lint....

**Comment by innerlee** [comment]
For the linting, try 

```bash
pre-commit install
pre-commit run --all-files
```

Follow the messages print by pre-commit, and fix them. And then commit the changes.

From the message below, you can see that isort is changing your codes, so you only need to commit them into the branch.

image

**Comment by HuangJunJie2017** [comment]
>From the message below, you can see that isort is changing your codes, so you only need to commit them into the branch.
where can i see how does the hook modify the codes? The modification has been done by the hook? where to check and commit the modifications?

**Comment by HuangJunJie2017** [comment]
emm, it seem that i should install these hook to modify my local code and then commit it?

**Pull request closed without merging**